File: aoc2015/python/day19.py
def parse(lines):
    rs = []
    for l in lines:
        rs.append(l.split(" => "))
    return rs

def replace(molecule, rep):
    a, b = rep
    parts = molecule.split(a)
    for i in range(len(parts) - 1):
        newm = parts[0]
        for j, p in enumerate(parts[1:]):
            if j == i:
                newm += b
            else:
                newm += a
            newm += p
        yield newm

def search(replacements, start):
    queue = [(start, 0)]
    seen = set()

    while queue:
        mol, steps = queue.pop()

        if mol == "e":
            return steps

        for r in replacements:
            for s in replace(mol, r):
                if s in seen:
                    continue
                seen.add(s)
                queue.append((s, steps + 1))
        queue.sort(key=lambda x: (len(x[0]), x[1], x[0]), reverse=True)

with open("day19.in") as fin:
    contents = fin.read().strip().split("\n")

medecine = contents.pop()
# pop blank line
contents.pop()
replacements = parse(contents)

molecules = set()
for r in replacements:
    for s in replace(medecine, r):
        molecules.add(s)

print(len(molecules))

# Reverse the replacements and work backwords from the large molecule to e
replacements = [(b, a) for (a, b) in replacements]
print(search(replacements, medecine))

File: aoc2015/src/bin/day19.rs
use std::fs::File;
use std::io::BufReader;
use std::io::BufRead;
use std::collections::HashSet;

struct Replacement {
    before: String,
    after: String,
}

fn replace(med: &str, rep: &Replacement) -> Vec<String> {
    let mut meds: Vec<String> = Vec::new();

    let parts: Vec<&str> = med.split(&rep.before).collect();
    for i in 1..parts.len() {
        let mut new_med = parts[0].to_string();
        for j in 1..parts.len() {
            if i == j {
                new_med = new_med + &rep.after;
            } else {
                new_med = new_med + &rep.before;
            }

            new_med = new_med + parts[j];
        }
        meds.push(new_med);
    }

    meds
}

fn search(med: &str, reps: &Vec<Replacement>) -> u32 {
    struct State {
        molecule: String,
        steps: u32
    }
    let mut queue: Vec<State> = Vec::new();
    queue.push(
        State{
            molecule: med.to_string(),
            steps: 0,
        }
    );

    while !queue.is_empty() {
        let s = queue.pop().unwrap();

        if s.molecule == "e" {
            return s.steps
        }

        for r in reps {
            for m in replace(&s.molecule, &r) {
                queue.push(
                    State{
                        molecule: m.clone(),
                        steps: s.steps + 1,
                    }
                );
            }
        }
        queue.sort_by(|a, b| b.molecule.len().cmp(&a.molecule.len()));
    }
    unreachable!();
}

fn main() {
    let f = File::open("day19.in")
        .ok()
        .expect("Error opening input");
    let file = BufReader::new(&f);

    let mut in_replacements = true;
    let mut replacements: Vec<Replacement> = Vec::new();
    let mut medecine: String = String::new();

    for l in file.lines() {
        let l = l.unwrap();
        if in_replacements {
            if l == "" {
                in_replacements = false;
            } else {
                let parts: Vec<&str> = l.trim().split(" => ").collect();
                replacements.push(Replacement{
                    before: parts[0].to_string(),
                    after: parts[1].to_string(),
                });
            }
        } else {
            medecine = l.trim().to_string();
        }
    }

    let mut meds = HashSet::new();
    for r in &replacements {
        for m in replace(&medecine, &r) {
            meds.insert(m);
        }
    }
    println!("{}", meds.len());

    let mut reversed: Vec<Replacement> = Vec::new();
    for r in &replacements {
        reversed.push(Replacement{
            before: r.after.clone(),
            after: r.before.clone(),
        });
    }
    let steps = search(&medecine, &reversed);
    println!("{}", steps)
}
